src/main/java/com/google/sps/servlets/ProfileServlet.java

```diff
@@ -106,7 +111,65 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     }
 
     @Override
-    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {  
+        String userId = userService.getCurrentUser().getUserId();
+        
+        // Don't update profile if cancel button is clicked
+        if (request.getParameter("submit").equals("Cancel")) {
+            response.sendRedirect("/profile.html?userID=" + userId);
+            return;
+        }
+        
+        //Make userId a query filter
+        Query query = new Query("User").setFilter(new Query.FilterPredicate("userId", Query.FilterOperator.EQUAL, userId));
+        PreparedQuery results = datastore.prepare(query);
+        Entity userEntity = results.asSingleEntity();
+
+        String role = (String) userEntity.getProperty("role"); 
+        String bio = Optional.ofNullable(request.getParameter("bio")).orElse("");
+        // Make list of selected topics, remove unchecked topics
+        List<Optional<String>> topics = new ArrayList<Optional<String>>();
+        topics.add(Optional.ofNullable(request.getParameter("math")));
+        topics.add(Optional.ofNullable(request.getParameter("english")));
+        topics.add(Optional.ofNullable(request.getParameter("other")));
+        List<String> topicsToStr = topics
+                                    .stream()
+                                    .filter(t -> t.isPresent())
+                                    .map(t -> t.get().toLowerCase())
+                                    .collect(Collectors.toList());
+
+        // User is a student, update their info
+        if (role.toLowerCase().equals("student")) {
+            Entity studentEntity = new Entity("Student");
+            updateStudentEntityAndPutInDatastore(datastore, studentEntity, userId, bio, topicsToStr);
+        } else {   // User is a tutor, update their info
+            Entity tutorEntity = new Entity("Tutor");
+            updateTutorEntityAndPutInDatastore(datastore, tutorEntity, userId, bio, topicsToStr);
+        }
+        response.sendRedirect("/profile.html?userID=" + userId);
+    }
+
+    /**
+    * Updates a student entity and puts it in datastore, used for testing
+    */
+    public void updateStudentEntityAndPutInDatastore(DatastoreService ds, Entity entity, String userId, String bio, List<String> topics) {
+        Query query = new Query("Student").setFilter(new Query.FilterPredicate("userId", Query.FilterOperator.EQUAL, userId));
+        PreparedQuery results = ds.prepare(query);
+        entity = results.asSingleEntity();
+        entity.setProperty("bio", bio);
```

Comment:
Should we add a try/catch here in case the query returns null?

Reply:
For some reason the compiler isn't accepting an EntityNotFoundException, so for now I'm just returning early if the query returns null.